Added section on  Manual (Reversed) Code Review for iOS (#1499)
OWASP/owasp-mstg: Document/0x06c-Reverse-Engineering-and-Tampering.md

![Objective-C runtime strings](Images/Chapters/0x06c/manual_reversing_ghidra_objc_runtime_strings.png "Objective-C runtime related strings")

On further careful analysis, we can spot the string, "Verification Failed", which is used in the pop-up when a wrong input is given. If you follow the **Xrefs** for this string, you will reach *buttonClick()* function in *ViewController* class. We will look into the *buttonClick()* function later in this section. On continuing checking the other strings in the application, only a few look a likely candidate for a 'hidden flag'. You can try them and verify as well. 

![Interesting strings in UnCrackable application](Images/Chapters/0x06c/manual_reversing_ghidra_strings.png "Interesting strings in UnCrackable application")

Moving forward, we have two paths to take. Either we can start analysing the *buttonClick()* function identified from the above step, or start analysing the application from the various entry points. In real world situation, most times you will be taking the first path, but from a learning perspective, in this section we will take the latter path. 

> An iOS application can have multiple entry points depending on at which particular state of the application lifecycle it is (documentation available at [Managing Your App's Life Cycle](https://developer.apple.com/documentation/uikit/app_and_environment/managing_your_app_s_life_cycle)). For example, when the application is started *[AppDelegate application:didFinishLaunchingWithOptions:]* is called, while *[AppDelegate applicationDidBecomeActive:]* is called when the application is moving from inactive to active state. Many applications execute critical code in these sections and is always a good starting point in order to follow the code systematically. 

Author: I am fine with

If we check the decompilation of this function, there are a few interesting things going on. For instance, there is a call to a native function at line 31 and a label is initialized with a `setHidden` flag set to 1 in lines 27-29. You can keep a note of these observations and continue exploring the other functions in this class. For brevity, exploring the other parts of the function is left as an exercise for the readers. 

In our first step, we observed that the application checks for the input string only when the UI button is pressed. Thus, analysing the *buttonClick()* function is an obvious target. As earlier mentioned, this function also contains the string we see in the pop-ups. At line 29 a decision is being made, which is based on the result of *isEqualString* (output saved in *uVar1* at line 23). The input for the comparison is coming from the text input field (from the user) and the value of the **label**!. The 'hidden flag' is stored in a label.  

![Decompilation of buttonClick function](Images/Chapters/0x06c/manual_reversing_ghidra_buttonclick_decompiled.png "Decompilation of buttonClick function")

> Now we have followed the complete flow and have all the information about the application flow. We also concluded that the 'hidden flag' is present in a label text and in order to determine the value of the label, we need to revisit *viewDidLoad()* function, and understand what is happening in the native function identified. Analysis of the native function is discussed in [Reviewing Disassembled Native Code](#Reviewing-Disassembled-Native-Code "Reviewing Disassembled Native Code").

Author: Sorry i keep messing up these formatting guidelines.

- In ARM64, a register is of 64 bit in size and referred to as **Xn**, where n is a number from 0 to 31. If lower (LSB) 32bits of the register is used then referred as **Wn**.
- The input parameters to a function are passed in **X0-X7** registers.
- The return value of the function is passed via **X0** register.
- Load (LDR) and store (STR) instructions are used to read or write to memory from/to a register.
- B, BL, BLX are branch instructions used for calling a function.

As mentioned above as well, Objective-C code is also compiled to native binary code, but analysing C/C++ native can be more challenging. In case of Objective-C there are various symbols (especially function names) present, which makes understanding of the code easy. From above section, presence of function names like *setText*, *isEqualStrings* helped us in quickly understanding the semantics of the code. In case of C/C++ native code, if all the debug symbols are stripped, there can be very few or no symbols present to assist in analyzing the binary. 

> Decompilers can help us in analyzing native code, but they should be used with caution. Modern decompilers are very sophisticated and among many techniques used by them to decompile code, few are heuristics based. Heuristics based techniques might not always give correct results, one such case being, determining the number of input parameters for a given native function. Having knowledge of analysing disassembled code, assisted with decompilers can make analyzing native code less error prone.  

Author: I want to avoid the use of only in "only a few....". Only gives an indication that we know about all the techniques and we are saying with certainty. Rather i want to keep it generic, and just state, there are heuristic based techniques in decompilers. do u feel same?

> Decompilers can help us in analyzing native code, but they should be used with caution. Modern decompilers are very sophisticated and among many techniques used by them to decompile code, few are heuristics based. Heuristics based techniques might not always give correct results, one such case being, determining the number of input parameters for a given native function. Having knowledge of analysing disassembled code, assisted with decompilers can make analyzing native code less error prone.  

Reviewer: I think that's fine. Thank you. I've corrected the suggestion accordingly :)

We will be analyzing the native function identified in *viewDidLoad()* function in the previous section. The function is at offset *0x1000080d4*. The return value of this function used in *setText* function call for the label. This text is used to compare against the user input. Thus, we can be sure that this function will be returning a string or equivalent. 

![Disassembly of the native function](Images/Chapters/0x06c/manual_reversing_ghidra_native_disassembly.png "Disassembly of the native function")

First thing we can see in the disassembly of the function is there is no input to the function. The registers X0-X7 are not read through out the function. Also, there are multiple calls to other functions like *0x100008158*, *0x10000dbf0* etc. 

Instructions to one such function call are below. Branch instruction *bl* is used to call the function *0x100008158*. 

```
1000080f0 1a 00 00 94     bl         FUN_100008158                
1000080f4 60 02 00 39     strb       w0,[x19]=>DAT_10000dbf0
```

> The return value from the function (return value in W0), is stored to the address in register X19 (*strb* stores a byte to the address in register). We can see the same pattern for other function calls, the returned value is stored in X19 register and each time the offset is one more than the previous function call. This behavior can be associated with populating each index of a string array at a time. Each return value is been written to an index of this string array. There are 11 such calls, and from the current evidence we can make an intelligent guess that length of the 'hidden flag' is 11. Towards the end of the disassembly, the function returns with the address to this string array. 

Reviewer: ```suggestion
The return value from the function (found in W0), is stored to the address in register X19 (`strb` stores a byte to the address in register). We can see the same pattern for other function calls, the returned value is stored in X19 register and each time the offset is one more than the previous function call. This behavior can be associated with populating each index of a string array at a time. Each return value is been written to an index of this string array. There are 11 such calls, and from the current evidence we can make an intelligent guess that length of the hidden flag is 11. Towards the end of the disassembly, the function returns with the address to this string array. 
```

```
100008148 e0 03 13 aa     mov        x0=>DAT_10000dbf0,x19
```

> To determine the value of the 'hidden flag' we need to know the return value of each of the subsequent function calls identified above. On analysing the function *0x100006fb4*, we can observe this function is much bigger and more complex than previous one we analysed. For analysing complex functions, function graphs can be very helpful, as it helps in understanding the control flow of the function. In Ghidra function graphs can be obtained by clicking **Display function graph** icon in the sub-menu.

Reviewer: ```suggestion
To determine the value of the hidden flag we need to know the return value of each of the subsequent function calls identified above. When analysing the function 0x100006fb4, we can observe that this function is much bigger and more complex than the previous one we analyzed. Function graphs can be very helpful when analyzing complex functions, as it helps into better understanding the control flow of the function. Function graphs can be obtained in Ghidra by clicking the **Display function graph** icon in the sub-menu.
```

![Function graph from 0x100006fb4](Images/Chapters/0x06c/manual_reversing_ghidra_function_graph.png)

> Manually analysing all the native functions completely will be time consuming and might not be the wisest approach. In such a scenario using dynamic analysis approach is recommended. By using the techniques like hooking or simply debugging the application, we can easily determine the returned values. In case dynamic analysis approach fails, then we can fallback and start manually analysing the functions. Dynamic analysis techniques are discussed in [Dynamic Analysis](#dynamic-analysis "Dynamic Analysis") section.  

Reviewer: I've introduced the idea of a feedback loop between static and dynamic. What do you think? 
```suggestion
Manually analysing all the native functions completely will be time consuming and might not be the wisest approach. In such a scenario using a dynamic analysis approach is highly recommended. For instance, by using the techniques like hooking or simply debugging the application, we can easily determine the returned values. Normally it's a good idea to use a dynamic analysis approach and then fallback to manually analysing the functions in a feedback loop. This way you can benefit from both approaches at the same time while saving time and reducing effort. Dynamic analysis techniques are discussed in "[Dynamic Analysis](#dynamic-analysis "Dynamic Analysis")" section.  
```

Author: yep. agreed.
